src/Analyzers/CSharp/Analyzers/UsePatternCombinators/CSharpUsePatternCombinatorsDiagnosticAnalyzer.cs

```diff
@@ -0,0 +1,126 @@
+                new LocalizableResourceString(nameof(CSharpAnalyzersResources.Use_pattern_matching), CSharpAnalyzersResources.ResourceManager, typeof(CSharpAnalyzersResources)))
+        {
+        }
+
+        protected override void InitializeWorker(AnalysisContext context)
+            => context.RegisterSyntaxNodeAction(AnalyzeNode,
+                SyntaxKind.LogicalAndExpression,
+                SyntaxKind.LogicalOrExpression,
+                SyntaxKind.LogicalNotExpression);
+
+        private void AnalyzeNode(SyntaxNodeAnalysisContext context)
+        {
+            var expression = (ExpressionSyntax)context.Node;
+
+            if (!IsTopmostExpression(expression))
+                return;
+
+            var syntaxTree = expression.SyntaxTree;
+            var cancellationToken = context.CancellationToken;
+
+            if (!((CSharpParseOptions)syntaxTree.Options).LanguageVersion.IsCSharp9OrAbove())
+                return;
+
+            var styleOption = context.Options.GetOption(CSharpCodeStyleOptions.PreferPatternMatching, syntaxTree, cancellationToken);
+            if (!styleOption.Value)
+                return;
+
+            var operation = context.SemanticModel.GetOperation(expression, cancellationToken);
+            if (operation is null)
+                return;
+
+            var pattern = CSharpUsePatternCombinatorsAnalyzer.Analyze(operation);
+            if (pattern is null)
+                return;
+
+            // Avoid rewriting trivial patterns, such as a single relational or a constant pattern.
+            if (IsTrivial(pattern))
+                return;
+
+            // C# 9.0 does not support pattern variables under `not` and `or` combinators.
+            if (HasIllegalPatternVariables(pattern))
+                return;
+
+            context.ReportDiagnostic(DiagnosticHelper.Create(
+                Descriptor,
+                expression.GetLocation(),
+                styleOption.Notification.Severity,
+                additionalLocations: null,
+                properties: null));
+        }
+
+        private static bool HasIllegalPatternVariables(AnalyzedPattern pattern, bool permitDesignations = true)
+        {
+            switch (pattern)
+            {
+                case Not p:
+                    return HasIllegalPatternVariables(p.Pattern, permitDesignations: false);
+                case Binary p:
+                    if (p.IsDisjunctive)
+                        permitDesignations = false;
+                    return HasIllegalPatternVariables(p.Left, permitDesignations) ||
+                           HasIllegalPatternVariables(p.Right, permitDesignations);
+                case Source p when !permitDesignations:
+                    return p.PatternSyntax.DescendantNodes()
+                        .OfType<SingleVariableDesignationSyntax>()
+                        .Any(variable => !variable.Identifier.IsMissing);
```

Comment:
hrmm... why the `missing` check? that seems unnecessary.

Reply:
This is exactly how the compiler check for this.

Reply:
cool.  in the fture, a comment to that effect couldbe helpful :)